sdk/eventhub/eventhubs-checkpointstore-blob/src/blobCheckpointStore.ts

```diff
@@ -253,21 +253,38 @@ export class BlobCheckpointStore implements CheckpointStore {
     blobName: string,
     metadata: OwnershipMetadata | CheckpointMetadata,
     etag: string | undefined
-  ) {
-    const blobClient = this._containerClient.getBlobClient(blobName);
+  ): Promise<BlobSetMetadataResponse> {
+    const blockBlobClient = this._containerClient.getBlobClient(blobName).getBlockBlobClient();
 
+    // When we have an etag, we know the blob existed.
+    // If we encounter an error we should fail.
     if (etag) {
-      return blobClient.setMetadata(metadata as Metadata, {
+      return blockBlobClient.setMetadata(metadata as Metadata, {
         conditions: {
           ifMatch: etag
         }
       });
     } else {
-      const blockBlobClient = blobClient.getBlockBlobClient();
+      try {
+        // Attempt to set metadata, and fallback to upload if the blob doesn't already exist.
+        // This avoids poor performance in storage accounts with soft-delete or blob versioning enabled.
+        // https://github.com/Azure/azure-sdk-for-js/issues/10132
+        return await blockBlobClient.setMetadata(metadata as Metadata);
+      } catch (err) {
+        // Check if the error is `BlobNotFound` and fallback to `upload` if it is.
+        if (err?.name !== "RestError") {
+          throw err;
+        }
+        const errorDetails = (err as RestError).details as { [field: string]: string } | undefined;
+        const errorCode = errorDetails?.errorCode ?? errorDetails?.Code;
```

Comment:
Checking `errorCode` may be enough.  `BlobNotFound` is responded in the header and gets into `details.errorCode`.  In the case where errors are from the response body, both `details.errorCode` and `details.Code` are set to the same error.

Reply:
Thanks for that info @jeremymeng, I wasn't sure which field was preferred. I can change to just check `errorCode` then.